src/macros-by-example.adoc

```diff
@@ -0,0 +1,827 @@
+
+[{mrk}]
+----
+(:byte_array 0 1 2 3 4 5 6 7 8) => [0, 1, 2, 3, 4, 5, 6, 7, 8]
+(:byte_array 9 -10 11)          => _**error**: -10 is not a valid uint8_
+(:byte_array 256)               => _**error**: 256 is not a valid uint8_
+----
+As above, Ion text doesn’t have syntax specifically denoting “8-bit unsigned integers”, so to
+keep text and binary capabilities aligned, the parser rejects invocations where an argument value
+exceeds the range of the binary-only type.
+
+The use of tagless types brings inherent tradeoffs which requires careful consideration, but in
+the right circumstances the density wins can be significant.
+
+
+=== Macro Shapes
+
+We can now introduce the final kind of input constraint, macro-shaped parameters.  To understand
+the motivation, consider modeling a scatter-plot as a list of points:
+
+----
+[{x:3, y:17}, {x:395, y:23}, {x:15, y:48}, {x:2023, y:5}, …]
+----
+
+Lists like these exhibit a lot of repetition. Since we already have a `point` macro, we can
+eliminate a fair amount:
+
+----
+[(:point 3 17), (:point 395 23), (:point 15 48), (:point 2023 5), …]
+----
+
+This eliminates all the ``x``s and ``y``s, but leaves repeated macro invocations.  We can try to
+wrap this in another macro, but we find the type constraints insufficient, since the tightest we
+can go is `*struct*`, and things aren’t really any better:
+
+[{mrk}]
+----
+(*public* scatterplot [(*struct\...* points)]
+  [points])
+----
+----
+(:scatterplot (:point 3 17) (:point 395 23) (:point 15 48) (:point 2023 5) …)
+----
+
+What we’d like is to build eliminate the `point` calls and just write a stream of pairs, something
+like:
+
+----
+(:scatterplot (3 17) (395 23) (15 48) (2023 5) …)
+----
+
+We can achieve exactly that with a macro-shaped parameter, in which we use the `point` macro as a
+pseudo-type:
+
+[{mrk}]
+----
+(*public* scatterplot [(*point\...* points)]
+  [points])
+----
+----
+(:scatterplot (3 17) (395 23) (15 48) (2023 5) …)
+=>
+[{x:3, y:17}, {x:395, y:23}, {x:15, y:48}, {x:2023, y:5}, …]
+----
+
+What’s happening here is that each argument S-expression like `(3 17)` is _implicitly an
+E-expression_ invoking the `point` macro.  The argument mirrors the shape of the inner macro,
+without repeating its name.  Further, expansion of the implied ``point``s happens automatically,
+so the overall behavior is just like the preceding struct-based variant and the `points`
+parameter produces a stream of structs.
+
+As the text notation here suggests, macro-shaped parameters are tagless, and the binary form also
+elides any opcodes mentioning `point` and just encodes its arguments with minimal delimiting.
+
+Macro types can be combined with any cardinality, following the same rules as before.  Note that
+when combined with `***` or `*+*` this requires callers to use two layers of delimiting
+containers, and outer delimiter for the parameter-stream as a whole, and an inner delimiter for
+each macro instance:
```

Comment:
> Where they have (:) they require an S-expression delimiting the implied invocation of foo_bar.

Okay, so these are equivalent, right?
```
(:ambiguous_macro (((:)) (3 4))) ⇒ [{}, {foo: 3, bar: 4}]
(:ambiguous_macro (() (3 4))) ⇒ [{}, {foo: 3, bar: 4}]
```

From earlier:
> Within the delimiting container, the invocation can have any number of values and/or macro
invocations.  The macro parameter produces the results of those expressions, concatenated into a
single stream, and the expander verifies that each value on that stream is acceptable by the
parameter’s declared type.

But when we are using macro-shape parameters, that goes out the window because the `()` of the implied macro invocation takes precedence, is that right?

Reply:
I don't think I'm going to be the only person who is confused by the way that delimiting containers and macro-shapes interact.

Would it be worth changing the syntax so that delimiting containers are `[]` and macro-shape delimiters are `()` so that you can tell at a glance which is which?

Reply:
Yeah, I think those are both correct.  In the former, the `(:)` is providing void to `f`, and optional `b` is implicitly void as well.

Reply:
> I don't think I'm going to be the only person who is confused by the way that delimiting containers and macro-shapes interact.
> 
> Would it be worth changing the syntax so that delimiting containers are `[]` and macro-shape delimiters are `()` so that you can tell at a glance which is which?

This is a *very* interesting suggestion to consider.  Allowing `()` for delimiters can avoid a bunch of commas, but as used in my own examples using `[]` makes for easier reading in many (if not most) cases.  I hadn't considered making the two concepts and notations one-to-one as you suggest, but I'll definitely look into it.